### PR TITLE
Add join-pushdown

### PR DESCRIPTION
**Pull request by yamatattsu**
Hi, Laurenz,

I did some cleanup the code. So I send this pull request.
This patch allow 2-way inner join SELECT query on a remote server.

Thanks,
Tatsuro Yamada

### CODE REVIEW

**Comment by laurenz** [comment]
I have committed 6328a31ca6906215679e43bc4ca7b6243c4573b3 which adds the necessary refactoring to make the code build with older PostgreSQL versions.

Most of it is mechanical stuff.
If you want, you can have a look if I broke anything, but the regression tests should show any problems once I get them done.

**Comment by laurenz** [comment]
I have added documentation with 732000bc75256e7ba7ee9f612be468d927dc1ad6. Did I forget anything important?

File: oracle_fdw.c [hunk 1]
```diff
@@ -903,101 +984,272 @@ oracleGetForeignPaths(PlannerInfo *root, RelOptInfo *baserel, Oid foreigntableid
...
+	{
+		ereport(DEBUG1, (errmsg("oracle_fdw: command type is not SELECT")));
+		return;
+	}
+
+	/*
+	 * N-way join is not supported, due to the column definition infrastracture.
+	 * If we can track relid mapping of join relations, we can support N-way join.
+	 */
+	if (outerrel->reloptkind != RELOPT_BASEREL || innerrel->reloptkind != RELOPT_BASEREL)
```
Comment (yamatattsu) [inline oracle_fdw.c]: I agree. It's good.

File: oracle_fdw.c [hunk 2]
```diff
@@ -903,101 +984,272 @@ oracleGetForeignPaths(PlannerInfo *root, RelOptInfo *baserel, Oid foreigntableid
...
+	else
+	{
+		/* There is no statistics, use 1000 as a default returned rows */
+		rows = 1000.0;
+	}
+
+	/* use outerrel's startup_cost and innerrels's startup_cost for joinrel's cost */
+	fdwState_o = (struct OracleFdwState *) outerrel->fdw_private;
+	fdwState_i = (struct OracleFdwState *) innerrel->fdw_private;
+	startup_cost = fdwState_o->startup_cost + fdwState_i->startup_cost;
```
Comment (yamatattsu) [inline oracle_fdw.c]: Okey. 
I understand that it is included cost of commnication between oracle and postgresql.

However, what do you think about how to calculate a cost against n-way join relations?
If we use this method which is joinrel's startup cost is based on sum of basrel's startup costs not fixed 10000, I suppose that it is easy to calculate when we will handle n-way join. 
We would be better to consider that cost model such as whether startup cost is always 10000 or not when implementing n-way join.

File: oracle_fdw.c [hunk 3]
```diff
@@ -903,101 +984,272 @@ oracleGetForeignPaths(PlannerInfo *root, RelOptInfo *baserel, Oid foreigntableid
...
+
+		if (has_trigger)
+		{
+			/* we need to fetch and return all columns */
+			for (i=0; i<fdwState->oraTable->ncols; ++i)
+				if (fdwState->oraTable->cols[i]->pgname)
+					fdwState->oraTable->cols[i]->used = 1;
+		}
+
+		local_exprs = fdwState->local_conds;
```
Comment (yamatattsu) [inline oracle_fdw.c]: I see.

File: oracle_fdw.c [hunk 4]
```diff
@@ -903,101 +984,272 @@ oracleGetForeignPaths(PlannerInfo *root, RelOptInfo *baserel, Oid foreigntableid
...
-		for (i=0; i<fdwState->oraTable->ncols; ++i)
-			if (fdwState->oraTable->cols[i]->pgname)
-				fdwState->oraTable->cols[i]->used = 1;
 	}
+	else
+	{	/* foreignrel->reloptkind == RELOPT_JOINREL */
 
-	/* create remote query */
-	fdwState->query = createQuery(fdwState, baserel, for_update, best_path->path.pathkeys);
-	elog(DEBUG1, "oracle_fdw: remote query is: %s", fdwState->query);
```
Comment (yamatattsu) [inline oracle_fdw.c]: Oops. Thanks.

File: oracle_fdw.c [hunk 5]
```diff
@@ -1025,7 +1277,10 @@ oracleExplainForeignScan(ForeignScanState *node, ExplainState *es)
 	char **plan;
 	int nrows, i;
 
-	elog(DEBUG1, "oracle_fdw: explain foreign table scan on %d", RelationGetRelid(node->ss.ss_currentRelation));
+	if (node->ss.ss_currentRelation)
+		elog(DEBUG1, "oracle_fdw: explain foreign table scan on %d", RelationGetRelid(node->ss.ss_currentRelation));
+	else
+		elog(DEBUG1, "oracle_fdw: explain foreign join");
```
Comment (yamatattsu) [inline oracle_fdw.c]: Agreed.

File: oracle_fdw.c [hunk 6]
```diff
@@ -2225,9 +2496,9 @@ struct OracleFdwState
 	fdwState->password = NULL;
 	fdwState->params = NIL;
 	fdwState->paramList = NULL;
-	fdwState->pushdown_clauses = NULL;
 	fdwState->temp_cxt = NULL;
 	fdwState->order_clause = NULL;
+	fdwState->where_clause = NULL;
```
Comment (yamatattsu) [inline oracle_fdw.c]: I see.

File: oracle_fdw.c [hunk 7]
```diff
@@ -2553,10 +2814,99 @@ char
...
+		/*
+		 * For a join relation FROM clause entry is deparsed as
+		 *
+		 * ((outer relation) <join type> (inner relation) ON (joinclauses)
+		 */
+		appendStringInfo(buf, "(%s %s JOIN %s ON ", join_sql_o.data,
+					   get_jointype_name(fdwState->jointype), join_sql_i.data);
+
+		/* Append join clause */
+		appendStringInfo(buf, "(");
```
Comment (yamatattsu) [inline oracle_fdw.c]: I understand.
It's better to remove unnecessary parentheses for readability.

File: oracle_fdw.c [hunk 8]
```diff
@@ -2553,10 +2814,99 @@ char
...
+
+	foreach(lc, exprs)
+	{
+		Expr  *expr = (Expr *) lfirst(lc);
+
+		/* Connect expressions with "AND" and parenthesize each condition. */
+		if (!is_first)
+			appendStringInfo(buf, " AND ");
+
+		appendStringInfo(buf, "(");
```
Comment (yamatattsu) [inline oracle_fdw.c]: Agreed.

File: oracle_fdw.c [hunk 9]
```diff
@@ -4254,29 +4627,47 @@ List
...
-	/* PostgreSQL table name */
-	result = lappend(result, serializeString(fdwState->oraTable->pgname));
-	/* number of columns in Oracle table */
-	result = lappend(result, serializeInt(fdwState->oraTable->ncols));
-	/* number of columns in PostgreSQL table */
-	result = lappend(result, serializeInt(fdwState->oraTable->npgcols));
-	/* column data */
-	for (i=0; i<fdwState->oraTable->ncols; ++i)
+
+	if(fdwState->oraTable)
```
Comment (yamatattsu) [inline oracle_fdw.c]: Oh, I realized it now.
Thanks for that.

File: oracle_fdw.c [hunk 10]
```diff
@@ -4599,6 +4989,7 @@ deparseTimestamp(Datum datum, bool hasTimezone)
 			datetime_tm.tm_min, datetime_tm.tm_sec, (int32)datetime_fsec,
 			(datetime_tm.tm_year > 0) ? "AD" : "BC");
 
+elog(DEBUG1, "deparseTimestamp s.data: %s", s.data);
```
Comment (yamatattsu) [inline oracle_fdw.c]: Yep. I forgot it to remove. Sorry. :)

File: oracle_fdw.c [hunk 11]
```diff
@@ -4642,9 +5033,298 @@ char
...
+
+	struct OracleFdwState *fdwState;
+	struct OracleFdwState *fdwState_o;
+	struct OracleFdwState *fdwState_i;
+
+	struct oraTable *oraTable_o;
+	struct oraTable *oraTable_i;
+
+	ListCell   *lc;
+	List       *joinclauses;
```
Comment (yamatattsu) [inline oracle_fdw.c]: I agree.
I'll add it on another occasion.

File: oracle_fdw.c [hunk 12]
```diff
@@ -4642,9 +5033,298 @@ char
...
+
+	/*
+	 * If either of the joining relations is marked as unsafe to pushdown, the
+	 * join can not be pushed down.
+	 */
+	fdwState = (struct OracleFdwState *) joinrel->fdw_private;
+	fdwState_o = (struct OracleFdwState *) outerrel->fdw_private;
+	fdwState_i = (struct OracleFdwState *) innerrel->fdw_private;
+	if (!fdwState_o || !fdwState_i)
+		return false;
```
Comment (yamatattsu) [inline oracle_fdw.c]: You are right. 
I see.

File: oracle_fdw.c [hunk 13]
```diff
@@ -4642,9 +5033,298 @@ char
...
+	 * required to be applied before joining the relations. Hence the join can
+	 * not be pushed down.
+	 */
+	if (fdwState_o->local_conds || fdwState_i->local_conds)
+		return false;
+
+	/* Separate restrict list into join quals and quals on join relation */
+
+	/* Only support INNER_JOIN */
+	if (jointype == JOIN_INNER){
```
Comment (yamatattsu) [inline oracle_fdw.c]: Agreed.

File: oracle_fdw.c [hunk 14]
```diff
@@ -4642,9 +5033,298 @@ char
...
+	 *
+	 * For a FULL OUTER JOIN, the other clauses from either relation can not
+	 * be added to the joinclauses or remote_conds, since each relation acts
+	 * as an outer relation for the other. Consider such full outer join as
+	 * unshippable because of the reasons mentioned above in this comment.
+	 *
+	 * The joining sides can not have local conditions, thus no need to test
+	 * shippability of the clauses being pulled up.
+	 */
+	switch (jointype)
```
Comment (yamatattsu) [inline oracle_fdw.c]: Agreed.

**Comment by yamatattsu** [comment]
>I have added documentation with 732000b. Did I forget anything important?

How about this? but I couldn't write naturally.
  * join with order by clauses pushed only join down.
What I want to say is that in case of combination of join-pushdown and sort-pushdown,
only allow join-pushdown.

**Comment by yamatattsu** [comment]
>Most of it is mechanical stuff.
>If you want, you can have a look if I broke anything, but the regression tests should show any problems >once I get them done.

I believe that your fix is right.
Just in case, I'm going to do regression test on next week. :-)

File: oracle_fdw.c (hunk 2 above)
Comment (laurenz) [inline oracle_fdw.c]: I don't know.
Will the startup cost be higher for an n-way join? Probably yes.
Will it be twice as high? Probably not.

My impression was that the 10000 are just so much hand-waving that it doesn't make much sense to overthink it, but we could also use something like `<inner table startup> + <outer table startup> - 9900`. Then the startup cost would be `10000 + 100 * <number of tables in the join>`.

That would somehow reflect that startup is probably slightly more expensive for a join query.

**Comment by laurenz** [comment]
>>    I have added documentation with 732000b. Did I forget anything important?
>
> How about this? but I couldn't write naturally.
>
> - join with order by clauses pushed only join down.
>   What I want to say is that in case of combination of join-pushdown and sort-pushdown,
>    only allow join-pushdown.

Good catch. I have added a line in e8a62501cec57a330fdc4ff1dd5bce13676f13df.

**Comment by yamatattsu** [comment]
Thanks!

2017-08-04 22:12 GMT+09:00 Laurenz Albe <notifications@github.com>:

**Pull request merged**